commcare-cloud/commcare_cloud/environment/schemas/postgresql.py

```diff
@@ -1,39 +1,102 @@
 import jsonobject
+import re
+
+import six
+
+from commcare_cloud.environment.constants import constants
+
+
+def alphanum_key(key):
+    """adapted from https://stackoverflow.com/a/2669120/240553"""
+    def convert(text):
+        return int(text) if text.isdigit() else text
+    return [convert(c) for c in re.split(r'([0-9]+)', key)]
 
 
 class PostgresqlConfig(jsonobject.JsonObject):
     _allow_dynamic_properties = False
 
     SEPARATE_SYNCLOGS_DB = jsonobject.BooleanProperty(default=True)
+    USE_PARTITIONED_DATABASE = jsonobject.BooleanProperty(default=False)
     DEFAULT_POSTGRESQL_HOST = jsonobject.StringProperty(default=None)
     DEFAULT_POSTGRESQL_USER = jsonobject.StringProperty(default="{{ secrets.POSTGRES_USERS.commcare.username }}")
     DEFAULT_POSTGRESQL_PASSWORD = jsonobject.StringProperty(default="{{ secrets.POSTGRES_USERS.commcare.password }}")
-
-    postgresql_dbs = jsonobject.ListProperty(lambda: DBOptions, required=True)
+    REPORTING_DATABASES = jsonobject.DictProperty(default=lambda: {"ucr": "ucr"})
+    dbs = jsonobject.ObjectProperty(lambda: SmartDBConfig)
 
     @classmethod
     def wrap(cls, data):
         self = super(PostgresqlConfig, cls).wrap(data)
-        for db in self.postgresql_dbs:
+        for db in self.generate_postgresql_dbs():
             if not db.user:
                 db.user = self.DEFAULT_POSTGRESQL_USER
             if not db.password:
                 db.password = self.DEFAULT_POSTGRESQL_PASSWORD
         return self
 
+    def to_json(self):
+        data = super(PostgresqlConfig, self).to_json()
+        data['postgresql_dbs'] = [db.to_json() for db in self.generate_postgresql_dbs()]
+        return data
+
     def replace_hosts(self, environment):
         if self.DEFAULT_POSTGRESQL_HOST is None:
             self.DEFAULT_POSTGRESQL_HOST = environment.groups['postgresql'][0]
         elif self.DEFAULT_POSTGRESQL_HOST != '127.0.0.1':
             self.DEFAULT_POSTGRESQL_HOST = environment.translate_host(
                 self.DEFAULT_POSTGRESQL_HOST, environment.paths.postgresql_yml)
 
-        for db in self.postgresql_dbs:
+        for db in self.generate_postgresql_dbs():
             if db.host is None:
                 db.host = self.DEFAULT_POSTGRESQL_HOST
             elif db.host != '127.0.0.1':
                 db.host = environment.translate_host(db.host, environment.paths.postgresql_yml)
 
+    def generate_postgresql_dbs(self):
+        return filter(None, [
+            self.dbs.main, self.dbs.synclogs,
```

Comment:
`main` could never be `None` right?

Reply:
Nope, it's `required=True`. If you specify `main` is null with

```
main:
```
or
```
main: null
```
then you'll get this error: `BadValueError: Property main is required.` If you (1) leave it out or (2) set it to `{}`
then it takes on the (recursive) default value (which is a `MainDBOptions` with all the default values).